I have following error, so I have to delete wait_for_update.
```
E       sqlalchemy.exc.NotSupportedError: (psycopg2.errors.FeatureNotSupported) FOR UPDATE cannot be applied to the nullable side of an outer join
E       
E       [SQL: SELECT dag.dag_id AS dag_dag_id, dag.root_dag_id AS dag_root_dag_id, dag.is_paused AS dag_is_paused, dag.is_subdag AS dag_is_subdag, dag.is_active AS dag_is_active, dag.last_scheduler_run AS dag_last_scheduler_run, dag.last_pickled AS dag_last_pickled, dag.last_expired AS dag_last_expired, dag.scheduler_lock AS dag_scheduler_lock, dag.pickle_id AS dag_pickle_id, dag.fileloc AS dag_fileloc, dag.owners AS dag_owners, dag.description AS dag_description, dag.default_view AS dag_default_view, dag.schedule_interval AS dag_schedule_interval, dag_tag_1.name AS dag_tag_1_name, dag_tag_1.dag_id AS dag_tag_1_dag_id 
E       FROM dag LEFT OUTER JOIN dag_tag AS dag_tag_1 ON dag.dag_id = dag_tag_1.dag_id 
E       WHERE dag.dag_id IN (%(dag_id_1)s, %(dag_id_2)s, %(dag_id_3)s, %(dag_id_4)s, %(dag_id_5)s, %(dag_id_6)s, %(dag_id_7)s, %(dag_id_8)s, %(dag_id_9)s, %(dag_id_10)s, %(dag_id_11)s, %(dag_id_12)s, %(dag_id_13)s, %(dag_id_14)s, %(dag_id_15)s, %(dag_id_16)s, %(dag_id_17)s, %(dag_id_18)s, %(dag_id_19)s, %(dag_id_20)s, %(dag_id_21)s, %(dag_id_22)s, %(dag_id_23)s, %(dag_id_24)s, %(dag_id_25)s, %(dag_id_26)s, %(dag_id_27)s, %(dag_id_28)s, %(dag_id_29)s, %(dag_id_30)s, %(dag_id_31)s, %(dag_id_32)s, %(dag_id_33)s, %(dag_id_34)s, %(dag_id_35)s, %(dag_id_36)s, %(dag_id_37)s, %(dag_id_38)s, %(dag_id_39)s, %(dag_id_40)s, %(dag_id_41)s, %(dag_id_42)s, %(dag_id_43)s, %(dag_id_44)s, %(dag_id_45)s, %(dag_id_46)s, %(dag_id_47)s, %(dag_id_48)s, %(dag_id_49)s, %(dag_id_50)s, %(dag_id_51)s, %(dag_id_52)s) FOR UPDATE]
E       [parameters: {'dag_id_1': 'test_on_kill', 'dag_id_2': 'example_external_task_marker_child', 'dag_id_3': 'test_utils', 'dag_id_4': 'example_xcom', 'dag_id_5': 'test_mark_success', 'dag_id_6': 'test_localtaskjob_double_trigger', 'dag_id_7': 'test_dagrun_states_deadlock', 'dag_id_8': 'example_branch_dop_operator_v3', 'dag_id_9': 'test_depends_on_past', 'dag_id_10': 'test_dagrun_states_fail', 'dag_id_11': 'impersonation_subdag.test_subdag_operation', 'dag_id_12': 'test_dagrun_states_root_fail', 'dag_id_13': 'clear_subdag_test_dag.daily_job', 'dag_id_14': 'test_dagrun_states_root_fail_unfinished', 'dag_id_15': 'example_passing_params_via_test_command', 'dag_id_16': 'latest_only_with_trigger', 'dag_id_17': 'example_subdag_operator.section-1', 'dag_id_18': 'test_start_date_scheduling', 'dag_id_19': 'test_backfill_pooled_task_dag', 'dag_id_20': 'latest_only', 'dag_id_21': 'test_dagrun_states_root_future', 'dag_id_22': 'test_task_start_date_scheduling', 'dag_id_23': 'example_kubernetes_executor', 'dag_id_24': 'test_latest_runs_1', 'dag_id_25': 'test_task_view_type_check', 'dag_id_26': 'example_trigger_target_dag', 'dag_id_27': 'impersonation_subdag', 'dag_id_28': 'test_default_impersonation', 'dag_id_29': 'test_dag_under_subdir2', 'dag_id_30': 'test_zip_dag', 'dag_id_31': 'test_retry_handling_job', 'dag_id_32': 'test_no_impersonation', 'dag_id_33': 'example_short_circuit_operator', 'dag_id_34': 'example_bash_operator', 'dag_id_35': 'test_run_ignores_all_dependencies', 'dag_id_36': 'test_subdag_operator', 'dag_id_37': 'example_subdag_operator.section-2', 'dag_id_38': 'example_kubernetes_executor_config', 'dag_id_39': 'example_branch_operator', 'dag_id_40': 'test_example_bash_operator', 'dag_id_41': 'example_trigger_controller_dag', 'dag_id_42': 'example_complex', 'dag_id_43': 'example_external_task_marker_parent', 'dag_id_44': 'test_subdag_operator.section-1', 'dag_id_45': 'tutorial', 'dag_id_46': 'example_subdag_operator', 'dag_id_47': 'test_heartbeat_failed_fast', 'dag_id_48': 'example_python_operator', 'dag_id_49': 'example_skip_dag', 'dag_id_50': 'test_impersonation', 'dag_id_51': 'clear_subdag_test_dag', 'dag_id_52': 'test_dagrun_states_success'}]
```

--- a/airflow/jobs/scheduler_job.py
+++ b/airflow/jobs/scheduler_job.py
@@ -821,8 +821,7 @@ class DagFileProcessor(LoggingMixin):
             return [], len(dagbag.import_errors)
 
         # Save individual DAGs in the ORM and update DagModel.last_scheduled_time
-        for dag in dagbag.dags.values():
-            dag.sync_to_db()
+        dagbag.sync_to_db()
 
         paused_dag_ids = {dag.dag_id for dag in dagbag.dags.values() if dag.is_paused}
 

--- a/airflow/models/dag.py
+++ b/airflow/models/dag.py
@@ -1483,6 +1483,7 @@ class DAG(BaseDag, LoggingMixin):
             .query(DagModel)
             .options(joinedload(DagModel.tags, innerjoin=False))
             .filter(DagModel.dag_id.in_(dag_ids))
+            .with_for_update(of=DagModel)
             .all()
         )
 

--- a/airflow/models/dag.py
+++ b/airflow/models/dag.py
@@ -26,14 +26,14 @@ import sys
 import traceback
 from collections import OrderedDict, defaultdict
 from datetime import datetime, timedelta
-from typing import Callable, Dict, FrozenSet, Iterable, List, Optional, Type, Union
+from typing import Callable, Collection, Dict, FrozenSet, Iterable, List, Optional, Type, Union
 
 import jinja2
 import pendulum
 from croniter import croniter
 from dateutil.relativedelta import relativedelta
 from sqlalchemy import Boolean, Column, ForeignKey, Index, Integer, String, Text, func, or_
-from sqlalchemy.orm import backref, relationship
+from sqlalchemy.orm import backref, joinedload, relationship
 from sqlalchemy.orm.session import Session
 
 from airflow import settings, utils
@@ -1455,63 +1455,91 @@ class DAG(BaseDag, LoggingMixin):
 
         return run
 
+    @classmethod
     @provide_session
-    def sync_to_db(self, owner=None, sync_time=None, session=None):
+    def bulk_sync_to_db(cls, dags: Collection["DAG"], sync_time=None, session=None):
+        """
+        Save attributes about list of DAG to the DB. Note that this method
+        can be called for both DAGs and SubDAGs. A SubDag is actually a
+        SubDagOperator.
+
+        :param dags: the DAG objects to save to the DB
+        :type dags: List[airflow.models.dag.DAG]
+        :param sync_time: The time that the DAG should be marked as sync'ed
+        :type sync_time: datetime
+        :return: None
+        """
+        if not dags:
+            return
+        from airflow.models.serialized_dag import SerializedDagModel
+
+        if sync_time is None:
+            sync_time = timezone.utcnow()
+        log.info("Sync %s DAGs", len(dags))
+        dag_by_ids = {dag.dag_id: dag for dag in dags}
+        dag_ids = set(dag_by_ids.keys())
+        orm_dags = (
+            session
+            .query(DagModel)
+            .options(joinedload(DagModel.tags, innerjoin=False))
+            .filter(DagModel.dag_id.in_(dag_ids))
+            .all()
+        )
+
+        existing_dag_ids = {orm_dag.dag_id for orm_dag in orm_dags}
+        missing_dag_ids = dag_ids.difference(existing_dag_ids)
+
+        for missing_dag_id in missing_dag_ids:
+            orm_dag = DagModel(dag_id=missing_dag_id)
+            dag = dag_by_ids[missing_dag_id]
+            if dag.is_paused_upon_creation is not None:
+                orm_dag.is_paused = dag.is_paused_upon_creation
+            log.info("Creating ORM DAG for %s", dag.dag_id)
+            session.add(orm_dag)
+            orm_dags.append(orm_dag)
+
+        for orm_dag in sorted(orm_dags, key=lambda d: d.dag_id):
+            dag = dag_by_ids[orm_dag.dag_id]
+            if dag.is_subdag:
+                orm_dag.is_subdag = True
+                orm_dag.fileloc = dag.parent_dag.fileloc  # type: ignore
+                orm_dag.root_dag_id = dag.parent_dag.dag_id  # type: ignore
+                orm_dag.owners = dag.parent_dag.owner  # type: ignore
+            else:
+                orm_dag.is_subdag = False
+                orm_dag.fileloc = dag.fileloc
+                orm_dag.owners = dag.owner
+            orm_dag.is_active = True
+            orm_dag.last_scheduler_run = sync_time
+            orm_dag.default_view = dag._default_view
+            orm_dag.description = dag.description
+            orm_dag.schedule_interval = dag.schedule_interval
+            orm_dag.tags = dag.get_dagtags(session=session)
+
+        session.commit()
+
+        for dag in dags:
+            cls.bulk_sync_to_db(dag.subdags, sync_time=sync_time, session=session)
+
+            if STORE_SERIALIZED_DAGS and not dag.is_subdag:
+                SerializedDagModel.write_dag(
+                    dag,
+                    min_update_interval=MIN_SERIALIZED_DAG_UPDATE_INTERVAL,
+                    session=session
+                )
+
+    @provide_session
+    def sync_to_db(self, sync_time=None, session=None):
         """
         Save attributes about this DAG to the DB. Note that this method
         can be called for both DAGs and SubDAGs. A SubDag is actually a
         SubDagOperator.
 
-        :param dag: the DAG object to save to the DB
-        :type dag: airflow.models.DAG
         :param sync_time: The time that the DAG should be marked as sync'ed
         :type sync_time: datetime
         :return: None
         """
-        from airflow.models.serialized_dag import SerializedDagModel
-
-        if owner is None:
-            owner = self.owner
-        if sync_time is None:
-            sync_time = timezone.utcnow()
-
-        orm_dag = session.query(
-            DagModel).filter(DagModel.dag_id == self.dag_id).first()
-        if not orm_dag:
-            orm_dag = DagModel(dag_id=self.dag_id)
-            if self.is_paused_upon_creation is not None:
-                orm_dag.is_paused = self.is_paused_upon_creation
-            self.log.info("Creating ORM DAG for %s", self.dag_id)
-            session.add(orm_dag)
-        if self.is_subdag:
-            orm_dag.is_subdag = True
-            orm_dag.fileloc = self.parent_dag.fileloc
-            orm_dag.root_dag_id = self.parent_dag.dag_id
-        else:
-            orm_dag.is_subdag = False
-            orm_dag.fileloc = self.fileloc
-        orm_dag.owners = owner
-        orm_dag.is_active = True
-        orm_dag.last_scheduler_run = sync_time
-        orm_dag.default_view = self._default_view
-        orm_dag.description = self.description
-        orm_dag.schedule_interval = self.schedule_interval
-        orm_dag.tags = self.get_dagtags(session=session)
-
-        session.commit()
-
-        for subdag in self.subdags:
-            subdag.sync_to_db(owner=owner, sync_time=sync_time, session=session)
-
-        # Write DAGs to serialized_dag table in DB.
-        # subdags are not written into serialized_dag, because they are not displayed
-        # in the DAG list on UI. They are included in the serialized parent DAG.
-        if STORE_SERIALIZED_DAGS and not self.is_subdag:
-            SerializedDagModel.write_dag(
-                self,
-                min_update_interval=MIN_SERIALIZED_DAG_UPDATE_INTERVAL,
-                session=session
-            )
+        self.bulk_sync_to_db([self], sync_time, session)
 
     @provide_session
     def get_dagtags(self, session=None):

--- a/airflow/models/dagbag.py
+++ b/airflow/models/dagbag.py
@@ -455,3 +455,10 @@ class DagBag(BaseDagBag, LoggingMixin):
             task_num=sum([o.task_num for o in stats]),
             table=pprinttable(stats),
         )
+
+    def sync_to_db(self):
+        """
+        Save attributes about list of DAG to the DB.
+        """
+        from airflow.models.dag import DAG
+        DAG.bulk_sync_to_db(self.dags.values())

--- a/airflow/utils/db.py
+++ b/airflow/utils/db.py
@@ -506,9 +506,9 @@ def initdb():
     create_default_connections()
 
     dagbag = DagBag()
-    # Save individual DAGs in the ORM
-    for dag in dagbag.dags.values():
-        dag.sync_to_db()
+    # Save DAGs in the ORM
+    dagbag.sync_to_db()
+
     # Deactivate the unknown ones
     DAG.deactivate_unknown_dags(dagbag.dags.keys())
 

--- a/tests/models/test_dag.py
+++ b/tests/models/test_dag.py
@@ -781,13 +781,14 @@ class TestDag(unittest.TestCase):
         )
         with dag:
             DummyOperator(task_id='task', owner='owner1')
+            subdag = DAG('dag.subtask', start_date=DEFAULT_DATE, )
+            # parent_dag and is_subdag was set by DagBag. We don't use DagBag, so this value is not set.
+            subdag.parent_dag = dag
+            subdag.is_subdag = True
             SubDagOperator(
                 task_id='subtask',
                 owner='owner2',
-                subdag=DAG(
-                    'dag.subtask',
-                    start_date=DEFAULT_DATE,
-                )
+                subdag=subdag
             )
         now = datetime.datetime.utcnow().replace(tzinfo=pendulum.timezone('UTC'))
         mock_now.return_value = now

--- a/tests/www/test_views.py
+++ b/tests/www/test_views.py
@@ -322,8 +322,7 @@ class TestAirflowBaseViews(TestBase):
     def setUpClass(cls):
         super().setUpClass()
         dagbag = models.DagBag(include_examples=True)
-        for dag in dagbag.dags.values():
-            dag.sync_to_db()
+        DAG.bulk_sync_to_db(dagbag.dags.values())
 
     def setUp(self):
         super().setUp()
@@ -1226,8 +1225,7 @@ class TestDagACLView(TestBase):
     def setUpClass(cls):
         super().setUpClass()
         dagbag = models.DagBag(include_examples=True)
-        for dag in dagbag.dags.values():
-            dag.sync_to_db()
+        DAG.bulk_sync_to_db(dagbag.dags.values())
 
     def cleanup_dagruns(self):
         DR = models.DagRun
@@ -2250,8 +2248,7 @@ class TestDecorators(TestBase):
     def setUpClass(cls):
         super().setUpClass()
         dagbag = models.DagBag(include_examples=True)
-        for dag in dagbag.dags.values():
-            dag.sync_to_db()
+        DAG.bulk_sync_to_db(dagbag.dags.values())
 
     def setUp(self):
         super().setUp()

